### PR TITLE
Glob patterns

### PR DESCRIPTION
**Pull request by JoeyZwicker**
- Adds glob patterns doc
- fixes a few links in fundamentals section
- update glob pattern references in other sections

@dwhitena please edit for everything -- spelling, grammar, format. :)

### CODE REVIEW

File: doc/fundamentals/distributed_computing.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
+# Distributed Computing
+
+Distributing computation across multiple workers is a fundamental part of processing any big data workload. There are two main questions to think about when trying to distribute computation:
```
Comment (dwhitena) [inline doc/fundamentals/distributed_computing.md]: Maybe "big data or computationally intensive workflow"

File: doc/fundamentals/distributed_computing.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+1) How many workers to spread computation across?
```
Comment (dwhitena) [inline doc/fundamentals/distributed_computing.md]: I think if you just just use `1.` and `2.` here instead of `1)` etc., it will be rendered as an actual numbered list.  Is that the same for `1)` etc.?

File: doc/fundamentals/distributed_computing.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+2) How to define which workers are responsible for which data?
+
+## Pachyderm Workers
+
+Before we dive into the above questions, there are a few details you should understand about Pachyderm workers. 
+
+Pachyderm workers are spun up when you create the pipeline and are left running in the cluster waiting for new jobs (data) to be available for processing (committed). This saves having to recreate and schedule the worker pod for every new job.
```
Comment (dwhitena) [inline doc/fundamentals/distributed_computing.md]: Is it enough to say "spun up"?  Will users know what that means?

Comment (dwhitena) [inline doc/fundamentals/distributed_computing.md]: Because in the next sentence you say "pod," and this might be the first place in our docs where we are really talking about pods.

File: doc/fundamentals/distributed_computing.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+Every worker for a given pipeline is an identical pod running the Docker image you specified in the [pipeline spec](../reference/pipline_spec.html). Your analysis code does not need do anything special to run in a distributed fashion. Instead, Pachyderm will spread out the data that needs to be processed across the various workers and make that data available for your code.
```
Comment (dwhitena) [inline doc/fundamentals/distributed_computing.md]: Ah, good here it is.  I would put this first before your "spun up" discussion.  Then it will make more sense.

File: doc/fundamentals/distributed_computing.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+Before we dive into the above questions, there are a few details you should understand about Pachyderm workers. 
+
+Pachyderm workers are spun up when you create the pipeline and are left running in the cluster waiting for new jobs (data) to be available for processing (committed). This saves having to recreate and schedule the worker pod for every new job.
+
+Every worker for a given pipeline is an identical pod running the Docker image you specified in the [pipeline spec](../reference/pipline_spec.html). Your analysis code does not need do anything special to run in a distributed fashion. Instead, Pachyderm will spread out the data that needs to be processed across the various workers and make that data available for your code. 
+
+## Controlling the Number of Workers (Parallelism)
+
+The number of workers that are used for a given pipeline is controlled by the `parallelism_spec` defined in the pipeline specification.
```
Comment (dwhitena) [inline doc/fundamentals/distributed_computing.md]: Link to the pipeline spec?

File: doc/fundamentals/distributed_computing.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    "constant": int        // if strategy == CONSTANT
+    "coefficient": double  // if strategy == COEFFICIENT
+```
+Pachyderm has two parallelism strategies: `CONSTANT` and `COEFFICIENT`.
+
+If you use the `CONSTANT` strategy, Pachyderm will start the number of workers that you specify. To use this strategy, set the field strategy to `CONSTANT`, and set the field `constant` to an integer value (e.g. `"constant":10` to use 10 workers).
+
+If you use the `COEFFICIENT` strategy, Pachyderm will start a number of workers that is a multiple of your Kubernetes cluster’s size. To use this strategy, set the field `coefficient` to a double. For example, if your Kubernetes cluster has 10 nodes, and you set `"coefficient": 0.5`, Pachyderm will start five workers. If you set it to 2.0, Pachyderm will start 20 workers (two per Kubernetes node).
+
+**NOTE:** The parallelism_spec is optional and will default to `“coefficient": 1”`, which means that it'll spawn one worker per Kubernetes node for this pipeline.
```
Comment (dwhitena) [inline doc/fundamentals/distributed_computing.md]: +1, thanks.

File: doc/fundamentals/distributed_computing.md [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+```
+Pachyderm has two parallelism strategies: `CONSTANT` and `COEFFICIENT`.
+
+If you use the `CONSTANT` strategy, Pachyderm will start the number of workers that you specify. To use this strategy, set the field strategy to `CONSTANT`, and set the field `constant` to an integer value (e.g. `"constant":10` to use 10 workers).
+
+If you use the `COEFFICIENT` strategy, Pachyderm will start a number of workers that is a multiple of your Kubernetes cluster’s size. To use this strategy, set the field `coefficient` to a double. For example, if your Kubernetes cluster has 10 nodes, and you set `"coefficient": 0.5`, Pachyderm will start five workers. If you set it to 2.0, Pachyderm will start 20 workers (two per Kubernetes node).
+
+**NOTE:** The parallelism_spec is optional and will default to `“coefficient": 1”`, which means that it'll spawn one worker per Kubernetes node for this pipeline. 
+
+## Sharding Data Across Workers (Glob Patterns)
```
Comment (dwhitena) [inline doc/fundamentals/distributed_computing.md]: This seems to be the first mention of "sharding."  It looks like we are wanting this to be connected with question number (2) above.  Maybe mention sharding up there or rephrase this to more specifically connect it with the above question.  Also, you might consider: 

1. internally linking each of the two questions above with the respective section below.
2. actually adding the (1) and (2) numbers to the section titles.

File: doc/fundamentals/distributed_computing.md [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+      "repo": {
+        "name": "string"
+      },
+      "glob": "string"
+    }
+  ]
+```
+That means you could easily define two inputs, one with the data highly distributed and another where it's grouped together. 
+
+### Datums
```
Comment (dwhitena) [inline doc/fundamentals/distributed_computing.md]: maybe add a brief statement above this section about why we are going to talk about datums etc.  Just to guide the reader.

File: doc/fundamentals/distributed_computing.md [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+Pachyderm uses the glob pattern to determine how many “datums” an input consists of. Datums are the unit of parallelism in Pachyderm. That is, Pachyderm attempts to process datums in parallel whenever possible.
+
+If you have two workers and define 2 datums, Pachyderm will send one datum to each worker. In a scenario where there are more datums than workers, Pachyderm will queue up extra datums and send them to workers as they finish processing previous datums. 
+
+### Glob Patterns
```
Comment (dwhitena) [inline doc/fundamentals/distributed_computing.md]: Maybe label this something like `Defining datums via glob patterns`?

File: doc/fundamentals/distributed_computing.md [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+   /San-Francisco.json
+   /Los-Angeles.json
+   ...
+/Colorado
+   /Denver.json
+   /Boulder.json
+   ...
+...
+```
+And you need to process all the data for a given state together, `/*` would also be the correct glob pattern. You'd have one datum per state, meaning all the cities for a given state would be processed together by a single worker, but each state can be processed independently.
```
Comment (dwhitena) [inline doc/fundamentals/distributed_computing.md]: "desired" instead of "correct"

File: doc/fundamentals/distributed_computing.md [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+   ...
+...
+```
+And you need to process all the data for a given state together, `/*` would also be the correct glob pattern. You'd have one datum per state, meaning all the cities for a given state would be processed together by a single worker, but each state can be processed independently. 
+
+If we instead used the glob pattern `/*/*` for the states example above, each `city.json` would be it's own datum. 
+
+Glob patterns also let you take only a particular directory (or subset of directories) as inputs instead of the whole input repo. 
+
+If we create a pipeline that is specifically only for California, we can use a glob pattern of `/California/*` to only use the data in that directory as input to our pipeline.
```
Comment (dwhitena) [inline doc/fundamentals/distributed_computing.md]: Combine with the sentence above.

File: doc/fundamentals/distributed_computing.md [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+```
+And you need to process all the data for a given state together, `/*` would also be the correct glob pattern. You'd have one datum per state, meaning all the cities for a given state would be processed together by a single worker, but each state can be processed independently. 
+
+If we instead used the glob pattern `/*/*` for the states example above, each `city.json` would be it's own datum. 
+
+Glob patterns also let you take only a particular directory (or subset of directories) as inputs instead of the whole input repo. 
+
+If we create a pipeline that is specifically only for California, we can use a glob pattern of `/California/*` to only use the data in that directory as input to our pipeline. 
+
+## Only Processing New Data
```
Comment (dwhitena) [inline doc/fundamentals/distributed_computing.md]: I kind of feel like this should go under the (2) question along with datums (i.e., `##` -> `###`). What do you think?

File: doc/fundamentals/distributed_computing.md (hunk 3 above)
Comment (JoeyZwicker) [inline doc/fundamentals/distributed_computing.md]: removed "pods". I think "spun up" is ok. Would you prefer "started"?

File: doc/fundamentals/distributed_computing.md (hunk 4 above)
Comment (JoeyZwicker) [inline doc/fundamentals/distributed_computing.md]: ok. reversed these paragraphs

File: doc/fundamentals/distributed_computing.md (hunk 7 above)
Comment (JoeyZwicker) [inline doc/fundamentals/distributed_computing.md]: changed "sharding" to "spreading" and added internal links

File: doc/fundamentals/distributed_computing.md (hunk 12 above)
Comment (JoeyZwicker) [inline doc/fundamentals/distributed_computing.md]: I'm on the fence, but I'm fine with that change. The only reason I can think of for keeping it as `##` is for discoverability because this is asked about a ton.

**Comment by JoeyZwicker** [comment]
@dwhitena Changes made.

**Comment by dwhitena** [comment]
LGTM.  Did you render and check the links?  Just a reminder.
